content/blog/2019-10-24-superblock-scheduling.md

```diff
@@ -0,0 +1,251 @@
+we do the following in a loop:
+- Make an bundle.
+- Take the instruction from the queue with the highest priority.
+- Add it to the bundle if compatible i.e. has no conflicts with the bundle.
+- Continue taking instructions from the queue until either the queue is empty or the instruction
+is incompatible with the current bundle.
+- For each element that we scheduled, check if their successors now have all their predecessors scheduled and add them to the queue
+
+The algorithm schedules a instruction only after all of its predecessors in the
+DAG have been scheduled. This means that to maintain program correctness,
+the predecessor relation must respect data dependencies.
+
+```
+v1: int = id x;
+x: int = const 10;
+```
+
+In the program above, we cannot reorder the instructions because the second
+instruction writes to `x` while the first reads from it. We can encode this
+constraint in the DAG by making the first instruction a predecessor of the second.
+
+### Generating Superblock
+
+The superblock algorithm generates a sequence of instruction and DAG that are
+used by the list scheduling algorithm to generate a program trace. This means
+that the superblock algorithm has to encode additional dependencies in the DAG
+to allow an unmodified list scheduling algorithm to work.
+
+The core insight with the superblock algorithm adding the DAG constraint that
+all the live out variables in a branch "read" from the conditional. This
+means that a write in a conditional cannot be moved outside it and in case
+a program exits from the middle of a trace, it will never observe the writes
+from a compacted branch body.
+
+For example,
+
+```
+  ...
+  v6: bool = gt v4 v5;
+  br v6 for.body.2 for.end.2;
+for.body.2:
+  v7: int = id result;
+  v8: int = id i;
+  v9: int = mul v7 v8;
+  result: int = id v9;
+  ...
+```
+
+When we consider this as a trace, we ignore the label and assume that the
+branch will jump to `for.body.2`. Without the constraint on the conditional,
+the following is valid reordering according to the list scheduling algorithm.
+
+```
+  ...
+  v6: bool = gt v4 v5;
+  v7: int = id result;
+  v8: int = id i;
+  v9: int = mul v7 v8;
+  result: int = id v9;
+  br v6 for.body.2 for.end.2;
+  ...
+```
+
+However, `v6` can be false which means we should have jumped to `for.end.2`
+instead but since the read was moved above the branch, `for.end.2` will observe
+the write to `result` which it shouldn't in the normal case.
+
+The other important ingredient in this algorithm is the heuristic that assigns priorities to instructions.
+We follow [Fisher's](https://people.eecs.berkeley.edu/~kubitron/courses/cs252-S12/handouts/papers/TraceScheduling.pdf)
+example and use the **highest levels first** heuristic. The priority of each node is the depth of the longest chain
+in the dependency DAG. He claims close to "optimal in practical environments" with this heuristic.
```

Comment:
It also occurs to me that this heuristic is an ingredient in the Gibbons instruction scheduling paper we read earlier in the semester.